src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matching;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    internal class DynamicControllerEndpointMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
```

Comment:
There are a few different design approaches I could have taken with MVC:

1. Add more nodes to the DFA stuff by doing an application model thing
2. Use the fallback endpoint's delegate to *chain* to an MVC endpoint
3. Use a matcher policy to substitute in another endpoint dynamically.

I rejected 1 because I don't really want this new routing feature to hang off of MVC options/conventions. Another idea that I could do would be to build this into the data sources themselves, which might be fine?

I rejected 2 because it has problems with things like authZ. You want the policies applied to the endpoint to be applied by middleware and that's only possible if endpoint is correct when middleware run.

I did 3 because I think we're **going to need this functionality**. I have another work item planned to implement features that people will use to port their `IRouter` code forward. What's enabled by this change is 100% needed for that.

-----

I'm looking for input on this because there are a few real tradeoffs here in terms of what's possible and how future enhancements will play out.

Reply:
Also @davidfowl I summon you.

Reply:
I believe you are right (as usual) in this.
1 would work but I understand why you don't want it.
2 makes sense.
3. I think I get it.

As someone not that familiar with this part of the codebase it would be great to get a more explicit pros/cons list. As it seems you clearly prefer 3. Is there any drawback to it?

Reply:
@JamesNK @pranavkm @davidfowl

Reply:
>  I have another work item planned to implement features that people will use to port their IRouter code forward. 

What sort of routing scenario would this help in?

Reply:
This is the typical CMS thing. Write an `IRouter` that treats the URL as a *slug* and then look that up in the database and put in the route values that go to the desired action.